include/gabia/secure_socket.hpp

```diff
@@ -0,0 +1,302 @@
+        encrypt_pdu(pdu, sequence);
+        asio::write(next_layer(), asio::buffer(pdu), error);
+        if (error) {
+            return 0;
+        }
+
+        return asio::buffer_size(sequence);
+    }
+
+    template <typename ConstBufferSequence>
+    auto write_some(const ConstBufferSequence& sequence) {
+        beast::error_code error = {};
+        auto n = write_some(sequence, error);
+        Expects(!error);
+        return n;
+    }
+
+    template <typename MutableBufferSequence>
+    std::size_t read_some(const MutableBufferSequence& sequence,
+                          beast::error_code& error) {
+        error = {};
+        if (decrypted_buffer.size() > 0) {
+            // We have already decrypted data ready to be retrieved from the
+            // buffer
+            auto consumed =
+                boost::asio::buffer_copy(sequence, decrypted_buffer.data());
+            decrypted_buffer.consume(consumed);
+            return consumed;
+        }
+
+        boost::endian::little_int16_t payload_length = 0;
+        size_t transferred = 0;
+        while (transferred < sizeof(payload_length) + payload_length +
+                                 crypto::aead_auth_tag_size) {
+            auto mb = raw_buffer.prepare(max_payload_length);
+            transferred += next_layer().read_some(mb, error);
+            if (error) {
+                return 0;
+            }
+            raw_buffer.commit(transferred);
+            extract_pdu_length(payload_length);
+        }
+
+        auto success = decrypt_pdu(payload_length);
+        if (!success) {
+            // Forged/spoofed message
+            // TODO: use a more appropriate error code
+            error = asio::error::no_permission;
+            decrypted_buffer.consume(payload_length);
+            return 0;
+        }
+
+        auto consumed =
+            boost::asio::buffer_copy(sequence, decrypted_buffer.data());
+        decrypted_buffer.consume(consumed);
+        return consumed;
+    }
+
+    template <typename MutableBufferSequence>
+    std::size_t read_some(const MutableBufferSequence& sequence) {
+        beast::error_code error = {};
+        auto n = read_some(sequence, error);
+        Expects(!error);
+        return n;
+    }
+
+    template <typename MutableBufferSequence, typename Handler>
+    auto async_read_some(MutableBufferSequence sequence, Handler&& handler) {
+        beast::async_completion<Handler, void(beast::error_code, size_t)> init{
+            handler};
+        if (decrypted_buffer.size() > 0) {
+            auto consumed =
+                boost::asio::buffer_copy(sequence, decrypted_buffer.data());
+            decrypted_buffer.consume(consumed);
+            next_layer().get_io_service().dispatch([
+                consumed, handler = std::move(init.completion_handler)
+            ]() mutable {
+                beast::error_code error{};
+                handler(error, consumed);
+            });
```

Comment:
To be honest I'm still a little bit confused by the design of these hooks, but I guess you just have to live with it... Oh and BTW, this would've worked just fine with strands, purely "by accident", because operator()() of a strand wrapped handler calls dispatch on the strand :D.

As a sidenote - do you know why asio uses these weird allocation hooks rather than std::allocator?